www/js/lib/uiUtil.js

```diff
@@ -43,34 +43,87 @@ define(rqDef, function(settingsStore) {
+            if (/^data:/i.test(content)) {
+                // Below uses Mozilla-recommended alternative to atob() : see https://developer.mozilla.org/en-US/docs/Glossary/Base64
+                content = util.dataURItoUint8Array(content);
+                // Or we could use webpHero's utility: it uses deprecated atob() but probably OK for image data
+                // content = webpHero.convertDataURIToBinary(content);
+            }
             // DEV: Note that webpMachine is single threaded and will reject an image if it is busy
             // However, the loadImagesJQuery() function in app.js is sequential (it waits for a callback
             // before processing another image) so we do not need to queue WebP images here
-            webpMachine.decode(content).then(function (uri) {
-                // DEV: WebpMachine.decode() returns a data: URI
-                // We callback before the node is set so that we don't incur slow DOM rewrites before processing more images
-                if (callback) callback(uri);
-                node.setAttribute(nodeAttribute, uri);
+            var canvas = document.createElement('canvas');
+            webpMachine.decodeToCanvas(canvas, content).then(function () {
+                if (callback) callback(); // Calling back as soon as possible speeds up extraction
+                if (params.useCanvasElements) {
+                    // Replace images by canvas. This is necessary for some browsers with obsolete anti-fingerprinting protection, like IceCat 60.7
+                    webpMachine.constructor.replaceImageWithCanvas(node, canvas);
+                } else {
+                    // For standard browsers that have access to the canvas, we simply convert the canvas to a data URI
+                    node.setAttribute(nodeAttribute, canvas.toDataURL());
+                }
             }).catch(function (err) {
                 console.error('There was an error decoding image in WebpMachine', err);
                 if (callback) callback();
             });
         } else {
-            var blob = new Blob([content], {
-                type: mimeType
-            });
-            var url = URL.createObjectURL(blob);
-            if (callback) callback(url);
-            node.addEventListener('load', function () {
-                URL.revokeObjectURL(url);
-            });
-            node.setAttribute(nodeAttribute, url);
+            if (callback) callback(); // Calling back as soon as possible speeds up extraction
+            // In browsers that support WebP natively, or for non-WebP images, we can simply convert the Uint8Array to a data URI
+            // DEV: we use FileReader method because btoa fails on utf8 strings (in SVGs, for example)
+            // See https://developer.mozilla.org/en-US/docs/Web/API/WindowBase64/Base64_encoding_and_decoding#The_Unicode_Problem
+            // This native browser method is very fast: see https://stackoverflow.com/a/66046176/9727685
```

Comment:
Sorry I should have been clearer. The revoking function actually concerns a Blob URL: [URL.revokeObjectURL()](https://developer.mozilla.org/en-US/docs/Web/API/URL/revokeObjectURL). Because we don't create a URL, there is nothing to revoke. A Blob itself is just a version of File and works like any other Object. There is no method for revoking an Object because it will be destroyed when the closure it was created inside ends.

Regarding the use of the FileReader method, I switched to using this in Kiwix JS Windows after I was getting errors with SVGs that could not be 64-bit encoded with `btoa()` (because SVGs can contain UTF8-encoded strings). Since we need to decode SVGs (the function here will run on any type of image), we either use this native browser method, or else we use the [Mozilla rewrite of `btoa()`](https://developer.mozilla.org/en-US/docs/Web/API/WindowBase64/Base64_encoding_and_decoding#The_Unicode_Problem).

Would you like me to add a timer and test both methods? The Mozilla method is a lot of code, but it would go in `util.js`. I haven't had any problems using the FileReader method in KJSW (it is an option currently that I've had maybe for 2 years or so).

Reply:
Thanks for the explanation.
If you already tested it on KJSW, it's fine: we can keep it